File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/security/DeliusUserDetailsServiceTest.java
package uk.gov.justice.digital.hmpps.oauth2server.security;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import uk.gov.justice.digital.hmpps.oauth2server.delius.model.DeliusUserPersonDetails;
import uk.gov.justice.digital.hmpps.oauth2server.delius.service.DeliusUserService;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;


@RunWith(MockitoJUnitRunner.class)
public class DeliusUserDetailsServiceTest {
    @Mock
    private DeliusUserService userService;

    private DeliusUserDetailsService service;

    @Before
    public void setup() {
        service = new DeliusUserDetailsService(userService);
    }

    @Test
    public void testHappyUserPath() {

        final var user = buildStandardUser("ITAG_USER");
        when(userService.getDeliusUserByUsername(user.getUsername())).thenReturn(Optional.of(user));

        final var itagUser = service.loadUserByUsername(user.getUsername());

        assertThat(itagUser).isNotNull();
        assertThat(itagUser.isAccountNonExpired()).isTrue();
        assertThat(itagUser.isAccountNonLocked()).isTrue();
        assertThat(itagUser.isCredentialsNonExpired()).isTrue();
        assertThat(itagUser.isEnabled()).isFalse();

        assertThat(((UserPersonDetails) itagUser).getName()).isEqualTo("Itag User");
    }

    @Test
    public void testUserNotFound() {

        when(userService.getDeliusUserByUsername(anyString())).thenReturn(Optional.empty());

        assertThatThrownBy(() -> service.loadUserByUsername("user")).isInstanceOf(UsernameNotFoundException.class);
    }

    private DeliusUserPersonDetails buildStandardUser(final String username) {
        return DeliusUserPersonDetails.builder()
                .username(username)
                .firstName("Itag")
                .enabled(false)
                .surname("User")
                .roles(List.of())
                .build();
    }
}

File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/security/DeliusUserDetailsServiceTest.kt
package uk.gov.justice.digital.hmpps.oauth2server.security

import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.anyString
import org.springframework.security.core.userdetails.UsernameNotFoundException
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User
import uk.gov.justice.digital.hmpps.oauth2server.delius.model.DeliusUserPersonDetails
import uk.gov.justice.digital.hmpps.oauth2server.delius.service.DeliusUserService
import java.util.*

class DeliusUserDetailsServiceTest {
  private val deliusUserService: DeliusUserService = mock()
  private val userService: UserService = mock()
  private val service = DeliusUserDetailsService(deliusUserService, userService)

  @Test
  fun `happy user path`() {
    val user = buildStandardUser()
    whenever(deliusUserService.getDeliusUserByUsername(anyString())).thenReturn(Optional.of(user))
    val itagUser = service.loadUserByUsername(user.username)
    assertThat(itagUser).isNotNull()
    assertThat(itagUser.isAccountNonExpired).isTrue()
    assertThat(itagUser.isAccountNonLocked).isTrue()
    assertThat(itagUser.isCredentialsNonExpired).isTrue()
    assertThat(itagUser.isEnabled).isFalse()
    assertThat((itagUser as UserPersonDetails).name).isEqualTo("Itag User")
  }

  @Test
  fun `user locked in auth`() {
    val user = buildStandardUser()
    whenever(deliusUserService.getDeliusUserByUsername(anyString())).thenReturn(Optional.of(user))
    whenever(userService.findUser(anyString())).thenReturn(Optional.of(User.builder().locked(true).build()))
    val itagUser = service.loadUserByUsername(user.username)
    assertThat(itagUser).isNotNull()
    assertThat(itagUser.isAccountNonExpired).isTrue()
    assertThat(itagUser.isAccountNonLocked).isFalse()
    assertThat(itagUser.isCredentialsNonExpired).isTrue()
    assertThat(itagUser.isEnabled).isFalse()
    assertThat((itagUser as UserPersonDetails).name).isEqualTo("Itag User")
  }

  @Test
  fun `user not found`() {
    whenever(deliusUserService.getDeliusUserByUsername(anyString())).thenReturn(Optional.empty())
    assertThatThrownBy { service.loadUserByUsername("user") }.isInstanceOf(UsernameNotFoundException::class.java)
  }

  private fun buildStandardUser(): DeliusUserPersonDetails =
      DeliusUserPersonDetails(username = "ITAG_USER", userId = "12345", firstName = "Itag", surname = "User", email = "a@b.com")
}
